### PR TITLE
feat: add support for twilio voice call

### PR DESCRIPTION
**Pull request by kshitij-k-osmosys**
### Pre-requisites

- [x] I have gone through the Contributing guidelines for [Submitting a Pull Request (PR)](../../CONTRIBUTING.md#submitting-a-pull-request-pr) and ensured that this is not a duplicate PR.
- [ ] I have performed preliminary testing using the [test suite](../../apps/api/OsmoX.postman_collection.json) to ensure that any existing features are not impacted and any new features are working as expected.
- [ ] I have updated the required [api docs](../../apps/api/docs/) as applicable.
- [ ] I have added/updated test cases to the [test suite](../../apps/api/OsmoX.postman_collection.json) as applicable

### PR Details

PR details have been updated as per the given format (see below)

- [x] PR title adheres to the format specified in guidelines (e.g., `feat: add admin login endpoint`)
- [x] Description has been added
- [x] Related changes have been added (optional)
- [ ] Screenshots have been added (optional)
- [x] Query request and response examples have been added (as applicable, in case added or updated)
- [ ] Documentation changes have been listed (as applicable)
- [ ] Test suite output is added (as applicable)
- [ ] Pending actions have been added (optional)
- [x] Any other additional notes have been added (optional)

### Additional Information

- [x] Appropriate label(s) have been added (`ready for review` should be added if the PR is ready to be reviewed)
- [x] Assignee(s) and reviewer(s) have been added (optional)

---

**Description:**

Add provider for making voice call via Twilio

**Related changes:**

- Create new Voice Call provider type
- Create new channel type VC_TWILIO for twilio voice calls
- Create base code for proper functionality for vc twilio
- Create migration file to add VC_TWILIO Master Provider
- From request object assumption is that we need to have atleast url or twiml object to form voice call message, so updated request valid decorator to ensure atleast one of them is sent in request body
- Update request dto
- Update request interface in service function
- Create a filter function in service to use the filtered object for twilio client call

**Screenshots:**

NA

**Query request and response:**

cURL

Note: `twiml` will not be read by twilioClient if `url` is present
```cURL
curl --location 'localhost:3000/notifications' \
--header 'x-api-key: OsmoX-test-key' \
--header 'Content-Type: application/json' \
--data '{
    
    "providerId": 10,
    "data": {
        "from": "+15005550006",
        "to": "+91xxxxxxxxxx",
        "url": "http://your-server-url/calls/message",
        "twiml": "<Response><Say>Hello there!</Say></Response>"
    }
}'
```

**Documentation changes:**

NA

**Test suite output:**

NA

**Pending actions:**

NA

**Additional notes:**
Reference Links

- https://www.twilio.com/en-us/blog/make-phone-call-using-node-js-twilio-programmable-voice
- https://www.twilio.com/docs/voice/quickstart/node#make-an-outgoing-phone-call-with-nodejs
- https://www.twilio.com/docs/voice/api/call-resource#call-status-values
- https://github.com/twilio/twilio-node/blob/main/src/rest/api/v2010/account/call.ts#L72

## Summary by CodeRabbit

- **New Features**
  - Added Twilio integration for voice call notifications.
  - Introduced new data validation for Twilio voice call parameters.
  - Implemented a service to handle Twilio voice call functionalities, including sending calls and retrieving delivery status.
  
- **Improvements**
  - Enhanced error handling and validation for Twilio voice call requests.
  
- **Configuration**
  - Added new constants for Twilio call statuses.

These changes enable more robust and reliable Twilio integration for voice call notifications, ensuring proper validation and error handling.

### CODE REVIEW

File: apps/api/src/modules/notifications/queues/queue.service.ts [hunk 1]
```diff
@@ -137,14 +142,24 @@ export class QueueService {
...
           await this.pushSnsNotificationConsumer.processPushSnsNotificationQueue(job.data.id);
           break;
+        // VC_TWILIO cases
+        case `${QueueAction.SEND}-${ChannelType.VC_TWILIO}`:
+          await this.vcTwilioNotificationsConsumer.processVcTwilioNotificationQueue(job.data.id);
+          break;
+        case `${QueueAction.DELIVERY_STATUS}-${ChannelType.VC_TWILIO}`:
+          await this.vcTwilioNotificationsConsumer.processVcTwilioNotificationConfirmationQueue(
+            job.data.id,
+          );
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/notifications/queues/queue.service.ts:162]: fixed in latest commit

File: apps/api/src/modules/notifications/dtos/providers/vcTwilio-data.dto.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+export class VcTwilioDataDto {
+  @IsNotEmpty()
+  from: string;
+
+  @IsNotEmpty()
+  to: string;
+
+  // Need atleast one of url, twiml for successful request
+  url?: string;
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/notifications/dtos/providers/vcTwilio-data.dto.ts]: Added IsOptional

File: apps/api/src/modules/providers/vc-twilio/vc-twilio.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+      providers: [VcTwilioService],
+    }).compile();
+
+    service = module.get<VcTwilioService>(VcTwilioService);
+  });
+
+  it('should be defined', () => {
+    expect(service).toBeDefined();
+  });
+});
```
[the remark this answers is not in the document]
Comment (LakshayaT) [reply apps/api/src/modules/providers/vc-twilio/vc-twilio.service.spec.ts:18]: This is an autogenerated file. We have a separate postman collection that we use for testing and don't see a need for modifying this file for now.

File: apps/api/src/modules/notifications/dtos/providers/vcTwilio-data.dto.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (LakshayaT) [reply apps/api/src/modules/notifications/dtos/providers/vcTwilio-data.dto.ts]: @kshitij-k-osmosys We should try to add a check to compulsorily have one or the other value as you mentioned in the comment. You can check if approach by @coderabbitai is fine
OR
You can have a look at this: https://github.com/OsmosysSoftware/osmo-x/blob/961d794fd64a78747a226ab3316a11687b9cf318/apps/api/src/modules/notifications/dtos/create-notification-attachment.dto.ts#L10 on how we did a similar check in the past

File: apps/api/src/database/migrations/1721088347000-VoiceCallSeedData.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    );
+  }
+
+  public async down(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`
+        DELETE FROM notify_master_providers
+        WHERE id IN (10);
+      `);
+  }
+}
```
[the remark this answers is not in the document]
Comment (LakshayaT) [reply apps/api/src/database/migrations/1721088347000-VoiceCallSeedData.ts:21]: We should be able to guarantee that the ID will be 10 for this new entry in the table. So this should be fine for now.

File: apps/api/src/modules/notifications/dtos/providers/vcTwilio-data.dto.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/notifications/dtos/providers/vcTwilio-data.dto.ts]: Added custom validator for checking one of url, twiml is in request body

File: apps/api/src/common/constants/notifications.ts [hunk 5]
```diff
@@ -53,6 +54,10 @@ export const ProviderDeliveryStatus = {
     SUCCESS_STATES: ['DELIVRD'],
     FAILURE_STATES: ['EXPIRED', 'UNDELIV', 'FAILED'],
   },
+  VC_TWILIO: {
+    SUCCESS_STATES: ['completed'],
+    FAILURE_STATES: ['queued', 'ringing', 'in-progress', 'canceled', 'busy', 'no-answer', 'failed'],
```
Comment (LakshayaT) [inline apps/api/src/common/constants/notifications.ts:59]: Are `'queued', 'ringing', 'in-progress'` really failure states? I would assume it means the request is still being processed/in progress.

File: apps/api/src/modules/notifications/dtos/providers/vcTwilio-data.dto.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+  @IsOptional()
+  twiml?: string;
+
+  @ValidateIf((obj: VcTwilioDataDto) => !obj.url && !obj.twiml)
+  @IsNotEmpty({
+    message: 'Request must include either a "url" or "twiml" parameter for Twilio voice calls',
+  })
+  validateUrlOrTwiml: string; // This is a dummy property to apply the custom validation logic
```
Comment (LakshayaT) [inline apps/api/src/modules/notifications/dtos/providers/vcTwilio-data.dto.ts:21]: Did this approach not work?: https://github.com/OsmosysSoftware/osmo-x/blob/961d794fd64a78747a226ab3316a11687b9cf318/apps/api/src/modules/notifications/dtos/create-notification-attachment.dto.ts#L10
Why did we need to add a dummy property?

File: apps/api/src/modules/providers/vc-twilio/vc-twilio.service.ts [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+      'callReason',
+      'callToken',
+      'recordingTrack',
+      'timeLimit',
+      'url',
+      'twiml',
+      'applicationSid',
+    ];
+
+    if (requestBody.url === undefined && requestBody.twiml === undefined) {
```
Comment (LakshayaT) [inline apps/api/src/modules/providers/vc-twilio/vc-twilio.service.ts:136]: Is this not redundant if we already have a check in DTO?

**Review by LakshayaT — COMMENTED** [review]
@kshitij-k-osmosys Can you please have a look at the queries below?

File: apps/api/src/modules/notifications/dtos/providers/vcTwilio-data.dto.ts (hunk 6 above)
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/notifications/dtos/providers/vcTwilio-data.dto.ts]: Removed dummy property. Used the approach suggested in create-notification-attachment.dto

File: apps/api/src/modules/providers/vc-twilio/vc-twilio.service.ts (hunk 7 above)
Comment (kshitij-k-osmosys) [reply apps/api/src/modules/providers/vc-twilio/vc-twilio.service.ts]: Removed this check

**Review by LakshayaT — APPROVED** [review]

**Pull request merged**